File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int widthOfBinaryTree(TreeNode* root) {
        int width = 0;
        queue<pair<TreeNode*,int>> que;
        que.push({root,0});

        while(que.size()!=0) {
            int size = que.size();
            int start = 0;
            int end = 0;
            for(int i=0;i<size;i++) {
                pair removeP = que.front(); que.pop();
                
                if(removeP.first->left) que.push({removeP.first->left,(long long)removeP.second*2+1});
                
                if(removeP.first->right) que.push({removeP.first->right,(long long)removeP.second*2+2});
                
                //starting & ending 
                if(i == 0) start = removeP.second;
                if(i == size-1) end = removeP.second;
            }
            width = max(width,end-start+1);
        }
        return width;
    }
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public class pair {
        int data;
        TreeNode node;
        
        pair(TreeNode node,int data) {
            this.node = node;
            this.data = data;
        }
    }
    public int widthOfBinaryTree(TreeNode root) {
        LinkedList<pair> que = new LinkedList<>();
        que.addLast(new pair(root,0));
        int ans = 0;
        while(que.size()!=0) {
            int startId=0;
            int endId = 0;
            int size = que.size();
            for(int i=0;i<size;i++) {
                pair removeP = que.removeFirst();
                if(removeP.node.left!=null)que.addLast(new pair(removeP.node.left,removeP.data*2+1));
                
                if(removeP.node.right!=null)que.addLast(new pair(removeP.node.right,removeP.data*2+2));
            
            if(i==0) startId = removeP.data;
            if(i==size-1) endId = removeP.data;
            }
            ans = Math.max(ans,endId-startId+1);
        }
        return ans;
    }
}